plugins/hls-change-type-signature-plugin/test/Main.hs

```diff
@@ -0,0 +1,138 @@
+        , testRegexes
+        , testTidyType
+    ]
+
+testRegexes :: TestTree
+testRegexes = testGroup "Regex Testing" [
+        testRegexOne
+        , testRegexTwo
+    ]
+    where
+        regex1 = errorMessageRegexes !! 0
+        regex2 = errorMessageRegexes !! 1
+
+testTidyType :: TestTree
+testTidyType = testGroup "Tidy Type Signature" [
+        tidyTypeSignatureTest "a -> a -> a" "a -> a -> a"
+        , tidyTypeSignatureTest "Int -> Double -> String" "Int -> Double -> String"
+        , tidyTypeSignatureTest "t1 a -> t2 c0 -> g7" "t a -> t0 b -> c"
+        , tidyTypeSignatureTest "t0 d -> t c" "t0 d -> t c"
+        , tidyTypeSignatureTest "a0 -> a1 -> a2" "a -> b -> c"
+        , tidyTypeSignatureTest "(a -> c1) -> c1" "(a -> b) -> b"
+        , tidyTypeSignatureTest "[a1] -> [b1] -> z" "[a] -> [b] -> z"
+        , tidyTypeSignatureTest "t0 a0 -> (a0 -> m0 b0) -> m0 (t0 b0)" "t a -> (a -> m b) -> m (t b)"
+    ]
+
+testRegexOne :: TestTree
+testRegexOne = testGroup "Regex One" [
+        regexTest "error1.txt" regex True
+        , regexTest "error2.txt" regex True
+        , regexTest "error3.txt" regex False
+        , regexTest "error4.txt" regex True
+    ]
+    where
+        regex = errorMessageRegexes !! 0
+
+testRegexTwo :: TestTree
+testRegexTwo = testGroup "Regex Two" [
+        regexTest "error1.txt" regex False
+        , regexTest "error2.txt" regex False
+        , regexTest "error3.txt" regex True
+        , regexTest "error4.txt" regex False
+    ]
+    where
+        regex = errorMessageRegexes !! 1
+
+testDataDir :: FilePath
+testDataDir = "test" </> "testdata"
+
+goldenChangeSignature :: FilePath -> (TextDocumentIdentifier -> Session ()) -> TestTree
+goldenChangeSignature fp = goldenWithHaskellDoc changeTypeSignaturePlugin (fp <> " (golden)") testDataDir fp "expected" "hs"
+
+codeActionTest :: FilePath -> Int -> Int -> TestTree
+codeActionTest fp line col = goldenChangeSignature fp $ \doc -> do
+    waitForDiagnostics  -- code actions are triggered from Diagnostics
+    actions <- getCodeActions doc (pointRange line col)
+    foundAction <- liftIO $ inspectCodeAction actions ["change signature"]
+    executeCodeAction foundAction
+
+codeActionProperties :: TestName -> [(Int, Int)] -> ([CodeAction] -> Session ()) -> TestTree
+codeActionProperties fp locs assertions = testCase fp $ do
+    runSessionWithServer changeTypeSignaturePlugin testDataDir $ do
+        openDoc (fp <.> ".hs") "haskell" >>= codeActionsFromLocs >>= findChangeTypeActions >>= assertions
+    where
+        codeActionsFromLocs doc = concat <$> mapM (getCodeActions doc . uncurry pointRange) locs
+
+findChangeTypeActions :: [Command |? CodeAction] -> Session [CodeAction]
+findChangeTypeActions = pure . filter isChangeTypeAction . rights . map toEither
+    where
+        isChangeTypeAction CodeAction{_kind} = case _kind of
+          Nothing -> False
+          Just kind -> case kind of
+            CodeActionQuickFix -> True
```

Comment:
FYI you can give your code action a kind "under" quickfix if you want, then you can match against it with https://hackage.haskell.org/package/lsp-types-1.4.0.1/docs/Language-LSP-Types.html#v:codeActionKindSubsumes

Reply:
Are you saying something like `CodeActionUnknown "quickfix.changeSignature"`

Reply:
Yeah (also it even has an `IsString` instance). There's no real logic to what to put as the bits underneath, use your judgement.